20210530-acme-zhiwei.md

```diff
@@ -0,0 +1,175 @@
+
+I propose to either use [acmez](https://pkg.go.dev/github.com/mholt/acmez) to implement ACME and write the solvers from scratch, or use the solvers in [CertMagic](https://github.com/caddyserver/certmagic) which already deal with the different challenges. Both are written in Go and fully-compliant with ACME.
+acmez is only for getting certificates, not managing them. Thus, we still have to write the logic for managing the certificates if we use purely acmez.
+
+ACMEZ can help us implement low level ACME protocol, but the major roadblock here would be to have a valid domain for which the certificate is issued. In order to perform ACME, the DNS server needs to own a valid domain.
+
+For example, if one owns a domain of `example.com`, this domain would be used to perform ACME challenges (HTTP/DNS) and allow us to issue certificate with required SANs (which could be `dns.example.com`). This subdomain of `dns.example.com` can be then used for serving DoH, i.e for all the subsquent queries the CoreDNS instance would use cert with SAN of `dns.example.com` to serve DoH. This certificate would be obtain using the original domain of `example.com`.
+
+The new configuration/Corefile for using acme for certificate management would be the following:
+```
+https://example.com {
+  tls {
+    acme <domain_name> {
+      challenge <challenge_type> /* (http01|dns01|tlsalpn)*/
+    }
+  }
+}
+```
+
+The above Corefile server block configures CoreDNS to serve DoH for all queries matching `example.com`. The `tls` block contains an `acme` label followed by a valid domain name. This domain name is used to perform ACME challenges and issue the certificate.
+
+#### Storage
+
+After successfully solving the ACME challenge, and recieving the required certificate from the CA, the next task is to store the certificate for further usage. Since, the certificate is to be used for serving DNS over HTTPS, it makes sense to store the certificate as [CERT](https://tools.ietf.org/html/rfc4398) Resource Record in the zone file. CERT RR provide a space in the DNS for certificates and related certificate revocation lists. Whenever a DNS (DoH) query hits the CoreDNS server, the server can fetch the required certificate from the zone file and use it to serve DoH. The certificates can be stored under a well-known domain (`dns.coredns.io`).
+
+Another plausible alternative to storing certificates is using TXT records. Instead of storing tls certificates as CERT RR, they can be stored as TXT Records under a well-known domain (`dns.coredns.io`) and can be used for serving DoH.
+
+### Clustered Environment
+
+In clustered environment, where there are multiple CoreDNS instances running, certificate sharing is a major concern. Certificate sharing is important as it is not feasible for each CoreDNS instance to fetch certificate individually.
+
+Since high availability is not a concern, a solution is to configure CoreDNS instances as "master" and "worker". 
```

Comment:
I am wondering if clustering can be split out into a separate plugin (`acme-master`/`acme-slave`/etc? Or other naming)? The acme might be part of the TLS though if users are not interested in clustering then they don't need to enable the plugin for clustering.

Reply:
I propose for it to stay under the `acme` plugin because it would reuse the code and logic in the case of a non-clustered environment.